src/Juvix/Compiler/Backend/Geb/Evaluator.hs

```diff
@@ -0,0 +1,377 @@
+  }
+
+makeLenses ''RunEvalArgs
+
+runEval :: RunEvalArgs -> Either JuvixError RunEvalResult
+runEval RunEvalArgs {..} =
+  case Geb.runParser _runEvalArgsInputFile _runEvalArgsContent of
+    Right (ExpressionMorphism m) -> do
+      let env :: Env =
+            Env
+              { _envEvaluatorOptions = _runEvalArgsEvaluatorOptions,
+                _envContext = Context.empty
+              }
+      if _runEvalArgsEvaluatorOptions ^. evaluatorOptionsNormalise
+        then RunEvalResultMorphism <$> nf' env m
+        else RunEvalResultGebValue <$> eval' env m
+    Right _ -> Left (error @JuvixError objNoEvalMsg)
+    Left err -> Left (JuvixError err)
+
+objNoEvalMsg :: Text
+objNoEvalMsg = "Geb objects cannot be evaluated, only morphisms."
+
+eval' :: Env -> Morphism -> Either JuvixError GebValue
+eval' env m =
+  run . runError $
+    mapError (JuvixError @EvalError) $
+      runReader env $
+        eval m
+
+nf' :: Env -> Morphism -> Either JuvixError Morphism
+nf' env m = run . runError $ runReader env (nf m)
+
+nf ::
+  Members '[Reader Env, Error JuvixError] r =>
+  Morphism ->
+  Sem r Morphism
+nf m = do
+  val :: GebValue <- mapError (JuvixError @EvalError) $ eval m
+  obj :: Object <-
+    runReader defaultInferenceEnv $
+      mapError (JuvixError @CheckingError) (inferObject m)
+  if
+      | requiresObjectInfo val -> quote (Just obj) val
+      | otherwise -> quote Nothing val
+
+type EvalEffects r = Members '[Reader Env, Error EvalError] r
+
+eval :: EvalEffects r => Morphism -> Sem r GebValue
+eval morph = do
+  -- env <- asks (^. envContext)
+  -- trace
+  --   ( "eval"
+  --       <> "\n"
+  --       <> "  arg:="
+  --       <> show morph
+  --       <> "\n"
+  --       <> "  env:="
+  --       <> show env
+  --       <> "\n"
+  --   ) $
+  case morph of
+    MorphismAbsurd x -> evalAbsurd x
+    MorphismApplication app -> evalApp app
+    MorphismBinop op -> evalBinop op
+    MorphismCase c -> evalCase c
+    MorphismFirst f -> evalFirst f
+    MorphismInteger i -> return $ GebValueMorphismInteger i
+    MorphismLambda l -> evalLambda l
+    MorphismLeft m -> GebValueMorphismLeft <$> eval m
+    MorphismPair p -> evalPair p
+    MorphismRight m -> GebValueMorphismRight <$> eval m
+    MorphismSecond s -> evalSecond s
+    MorphismUnit -> return GebValueMorphismUnit
+    MorphismVar x -> evalVar x
+
+evalVar :: EvalEffects r => Var -> Sem r GebValue
+evalVar var = do
+  ctx <- asks (^. envContext)
+  let val = Context.lookup (var ^. varIndex) ctx
+  -- trace ("varLookup := " <> show val) $
+  return val
+
+evalAbsurd :: EvalEffects r => Morphism -> Sem r GebValue
+evalAbsurd morph =
+  throw
+    EvalError
+      { _evalErrorMsg = "Absurd can not be evaluated.",
+        _evalErrorGebValue = Nothing,
+        _evalErrorGebExpression = Just morph
+      }
+
+evalPair :: EvalEffects r => Pair -> Sem r GebValue
+evalPair pair = do
+  left <- eval $ pair ^. pairLeft
+  right <- eval $ pair ^. pairRight
```

Comment:
If you want call-by-value, you also need to force evaluation of `left` and `right` here. Analogously for all other constructors and primitives.

Reply:
Done.